src/mono/wasm/runtime/startup.ts

```diff
@@ -57,26 +57,36 @@ export function mono_wasm_set_runtime_options(options: string[]): void {
 }
 
 async function _fetch_asset(url: string): Promise<Response> {
-    if (typeof (fetch) === "function") {
-        return fetch(url, { credentials: "same-origin" });
-    }
-    else if (ENVIRONMENT_IS_NODE) {
-        // eslint-disable-next-line @typescript-eslint/no-var-requires
-        const fs = require("fs");
-        const arrayBuffer = await fs.readFileSync(url);
-        return <Response><any> {
-            ok: true,
-            url,
-            arrayBuffer: () => arrayBuffer,
-            json: () => JSON.parse(arrayBuffer)
-        };
+    try {
+        if (typeof (fetch) === "function") {
+            return fetch(url, { credentials: "same-origin" });
+        }
+        else if (ENVIRONMENT_IS_NODE) {
+            // eslint-disable-next-line @typescript-eslint/no-var-requires
+            const fs = require("fs");
+            const arrayBuffer = await fs.readFileSync(url);
+            return <Response><any> {
+                ok: true,
+                url,
+                arrayBuffer: () => arrayBuffer,
+                json: () => JSON.parse(arrayBuffer)
+            };
+        }
+        else if (typeof (read) === "function") {
+            return <Response><any> {
+                ok: true,
+                url,
+                arrayBuffer: () => new Uint8Array(read(url, "binary")),
+                json: () => JSON.parse(read(url))
+            };
+        }
     }
-    else if (typeof (read) === "function") {
+    catch (e) {
         return <Response><any> {
-            ok: true,
+            ok: false,
```

Comment:
Maybe store the error as a property so the caller can find out what the error was without an extra throw+catch